### ISSUE TITLE
Error: knexStringcase is not a function

### ISSUE THREAD

**Issue by arpowers** [Kequc/knex-stringcase#27]
Looks like there is a type-error as the "default" import from "knex-stringcase" isn't a function... 

After looking at the code, it appears it should be a named export? 

anyway, 
```ts
import knexStringcase from 'knex-stringcase'
// error knexStringcase is not a function
const opts = knexStringcase(knexOptions) 
```

**Comment by arpowers** [comment]
I would have bumped this to `2.0.0` as it introduces a major feature ;)

**Comment by Kequc** [comment]
That definitely doesn't sound good, the following code works in a quick test typescript project I made.

```
import knex from 'knex';
import knexStringcase from 'knex-stringcase';

const options = knexStringcase({ client: 'pg' });
const db = knex(options);
```

But all the tooling changed. Are you using typescript, or is it esm, or?

**Comment by Kequc** [comment]
I tried the same using esm and that works. If anyone knows what could cause the above problem please let me know.

**Comment by arpowers** [comment]
It may be related to module resolution that is used, here is my relevant tsconfig: 

```
{
  "compilerOptions": {
    "target": "ESNext", 
    "lib": ["ESNext", "DOM", "DOM.Iterable"],
    "skipLibCheck": true,  
    "composite": true, 
    "module": "Preserve",
    "verbatimModuleSyntax": true,
    "noImplicitOverride": true,  
    "strict": true,
    "noEmit": true,
    "outDir": "dist",
    "removeComments": true,
    "esModuleInterop": true, 
    "skipDefaultLibCheck": true,
    "incremental": true, 
    "moduleDetection": "force" 
  }
 ```

**Comment by Kequc** [comment]
With your suggested tsconfig file, which was a good idea, I still have a functioning implementation.
```
{
    "compilerOptions": {
        "target": "ESNext",
        "lib": [
            "ESNext",
            "DOM",
            "DOM.Iterable"
        ],
        "skipLibCheck": true,
        "composite": true,
        "module": "Preserve",
        "verbatimModuleSyntax": true,
        "noImplicitOverride": true,
        "strict": true,
        "noEmit": true,
        "outDir": "dist",
        "removeComments": true,
        "esModuleInterop": true,
        "skipDefaultLibCheck": true,
        "incremental": true,
        "moduleDetection": "force"
    }
}
```

**Comment by arpowers** [comment]
Ok! I'll investigate more tomorrow, thanks for working on this.

**Comment by Kequc** [comment]
Did you used to have a typescript stub for knex-stringcase, and recently updated to the typescript module? Try deleting the stub.

**Comment by arpowers** [comment]
Ok did a bit more debugging, in my case, what is happening is that its importing with a `default` name on the export: 

```ts
import knexStringcase from 'knex-stringcase'

// knexStringcase.default = Function
```

I got it working just using an async import: 

```ts
    const knexStringcase = await import('knex-stringcase')
    const opts: Knex.Config = knexStringcase.default(knexOptions as KnexStringcaseConfig) as Knex.Config

```

this problem is resolved for me, but it may still be a bug in some sense, close at your discretion. And THANK YOU for the plugin!

**Comment by Kequc** [comment]
This seems odd. I imagine you don't have to do all that when you import knex so what's the difference with my library? I'll have to do some more research.

Hopefully this is isolated to only a couple users and if anyone knows what's going on let us know.

**Comment by Kequc** [comment]
Alright, I talked to chatgpt for about half an hour. Here is the significant outcome of debugging this issue. It seems like likely the reason why your tsconfig worked for me and not for you is because I am using a newer version of node 21.4. That might be the case I'm not sure.

Regardless I've updated the tsconfig and build for the library using all of the recommended functionality for compatibility and you can try version 1.5.4, if that doesn't work consider what chatgpt suggests below. Or if you don't want to then the recommended import is one of the following options.

```
const knexStringcase - require('knex-stringcase');
const knexStringcase = require('knex-stringcase').default;
import knexStringcase from 'knex-stringcase';
import { default as knexStringcase } from 'knex-stringcase';
```

Although, I consider a lot of those to be hacks.

That upset within the javascript community regarding difficulties with innteroperability between commonjs and esm might have hit here after I modernized the codebase (as best I could).

ChatGPT's advice:

> ### The Actual Cause: `module: "Preserve"`
> 
> The key detail here is `"module": "Preserve"` in their `tsconfig.json`. This option tells TypeScript **not** to transform the module syntax in the output files, preserving the original import/export syntax from the source code. In other words, the compiled TypeScript code keeps ESModule-style imports and exports as they are, rather than transforming them for a CommonJS environment.
> 
> Here’s what’s likely happening in this scenario:
> 
> 1. **Module Preservation**: When `"module": "Preserve"` is set, TypeScript doesn't emit any changes to how imports/exports are handled, so if the environment that executes the compiled code is CommonJS (like Node.js), it expects `require()` to work like a CommonJS module.
>    
> 2. **How the Environment Executes the Code**: If the execution environment is CommonJS, but the module syntax is preserved (i.e., ESModule-style exports), then `require()` will see the default export under the `.default` property because CommonJS doesn't natively support default exports the way ESModules do. This mismatch is what forces the user to access `.default`.
> 
> 3. **Interplay with `esModuleInterop`**: While `esModuleInterop: true` helps to smooth over the differences between CommonJS and ESModules, it doesn't help if the module system is **preserved** (i.e., left untouched). In that case, the runtime environment needs to support ESModules natively, or you get this issue with `.default` being required.
> 
> ### Why It Worked in Your Case:
> When you tested your application, you likely had a different setting for `"module"` (such as `"commonjs"` or `"esnext"`), which allows TypeScript to properly transform the ESModule syntax or target an environment that handles it natively.
> 
> ### Solution for the User:
> 
> If the user wants to avoid having to use `.default`, they should change their `tsconfig.json` from:
> 
> ```json
> "module": "Preserve"
> ```
> 
> to:
> 
> ```json
> "module": "ESNext"  // or "CommonJS", depending on the environment
> ```
> 
> This allows TypeScript to either:
> - Target an ESModule environment properly (with `"ESNext"`), or
> - Transform the module syntax for CommonJS (`"CommonJS"`), where the interop layer will work as expected.
> 

Thanks for bringing my attention to this!

**Comment by arpowers** [comment]
@Kequc further info for intellectual reasons, here is the ACTUAL JS that is being used in my install: 

as you can see, the only export is on the 'default' keyword.. and it has some interop code...  Im not sure what the difference is between this and the old package; but anyway, the typical thing you'd see is an ESM build and a CJS build these days to make this more predictable... 

```js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = knexStringcase;
const converter_factory_1 = __importDefault(require("./converter-factory"));
const key_converter_factory_1 = __importDefault(require("./key-converter-factory"));
function knexStringcase(config) {
    var _a, _b;
    const options = Object.assign({}, config);
    delete options.appWrapIdentifier;
    delete options.appPostProcessResponse;
    delete options.appStringcase;
    delete options.stringcase;
    delete options.recursiveStringcase;
    options.wrapIdentifier = wrapIdentifierFactory((0, converter_factory_1.default)((_a = config === null || config === void 0 ? void 0 : config.stringcase) !== null && _a !== void 0 ? _a : 'snakecase'), config === null || config === void 0 ? void 0 : config.appWrapIdentifier, config === null || config === void 0 ? void 0 : config.wrapIdentifier);
    options.postProcessResponse = postProcessResponseFactory((0, key_converter_factory_1.default)((0, converter_factory_1.default)((_b = config === null || config === void 0 ? void 0 : config.appStringcase) !== null && _b !== void 0 ? _b : 'camelcase'), config === null || config === void 0 ? void 0 : config.recursiveStringcase), config === null || config === void 0 ? void 0 : config.postProcessResponse, config === null || config === void 0 ? void 0 : config.appPostProcessResponse);
    return options;
}
function wrapIdentifierFactory(converter, before, after) {
    return function wrapIdentifier(value, origImpl, queryContext) {
        let output = value;
        if (typeof before === 'function') {
            output = before(output, queryContext);
        }
        output = converter(output);
        if (typeof after === 'function') {
            output = after(output, origImpl, queryContext);
        }
        else {
            output = origImpl(output);
        }
        return output;
    };
}
function postProcessResponseFactory(keyConverter, before, after) {
    return function postProcessResponse(result, queryContext) {
        let output = result;
        if (typeof before === 'function') {
            output = before(output, queryContext);
        }
        output = keyConverter(output, 'root', queryContext);
        if (typeof after === 'function') {
            output = after(output, queryContext);
        }
        return output;
    };
}
```

**Comment by Kequc** [comment]
As far as I understand it there's a lot of tricky business making exports and imports compatible across various implementations. Version 1.4.x didn't have a compile step. So I just said `module.exports = knexStringcase`, which worked as a default export, and typescript was good at importing it as a default import.

I don't know the technical reasons behind it all but it seems like newer javascript there's a lot of different ways to make it work. And what I'm using seems to be the modern standard.

ChatGPT said something about making two different versions. But, I dunno about all that. Is it worth it?

**Comment by arpowers** [comment]
@Kequc its worth it as a learning exercise for making production grade packages... If i recall its just a few lines of code or setting some options somewhere. Hopefully the whole cjs/mjs compatibility nightmare is behind us!

**Comment by VividVisions** [comment]
FYI: I just installed 1.5.5 via npm, imported it into my non-TS, ESM-only project with 

```js
import knexStringcase from 'knex-stringcase';
```

and got the "knexStringcase  is not a function" error because the function is exported under knexStringcase.default. Something's still not right here.

**Comment by Kequc** [comment]
First, I'd like to apologize for the ongoing problems with commonjs/esm interoperability. I've published version 1.6 which contains both versions, one for commonjs and one for esm.

The library itself has also been updated to use the latest node and type "module" (no tsx necessary when developing locally). Hopefully, it's both a modern library and one that supports the most number of different environments cleanly. I'll be very disappointed otherwise.
